File: benchmarks/compare/pipeline/pipeline.go
// pipeline.go — range -> map -> filter -> reduce, materializing each stage into
// an explicit slice (no lazy fusion). Prints "RESULT=<int>".
package main

import "fmt"

const n = 2000000

func main() {
	a := make([]int64, n)
	for i := int64(0); i < n; i++ {
		a[i] = i
	}

	b := make([]int64, len(a))
	for i, x := range a {
		b[i] = x * 2
	}

	c := make([]int64, 0, len(b))
	for _, x := range b {
		if x%3 == 0 {
			c = append(c, x)
		}
	}

	var total int64 = 0
	for _, x := range c {
		total += x
	}
	fmt.Printf("RESULT=%d\n", total)
}

File: benchmarks/compare/pipeline/pipeline.rs
// pipeline.rs — range -> map -> filter -> reduce, materializing each stage into
// an explicit Vec (no lazy iterator fusion). Prints "RESULT=<int>".
const N: i64 = 20_000_000;

fn main() {
    let a: Vec<i64> = (0..N).collect();
    let b: Vec<i64> = a.iter().map(|x| x * 2).collect();
    let c: Vec<i64> = b.iter().cloned().filter(|x| x % 3 == 0).collect();
    let mut total: i64 = 0;
    for x in &c {
        total += x;
    }
    println!("RESULT={}", total);
}
